packages/smart-contracts/src/contracts/ERC20TransferrableReceivable.sol

```diff
@@ -0,0 +1,170 @@
+
+  address public paymentProxy;
+
+  // Event to declare payments to a receivableTokenId
+  event Payment(
+    address sender,
+    address recipient,
+    address paymentProxy,
+    uint256 receivableTokenId,
+    address tokenAddress,
+    uint256 paymentId,
+    bytes indexed paymentReference
+  );
+
+  // Event to declare a transfer with a reference
+  // This event is emitted from a delegatecall to an ERC20FeeProxy contract
+  event TransferWithReferenceAndFee(
+    address tokenAddress,
+    address to,
+    uint256 amount,
+    bytes indexed paymentReference,
+    uint256 feeAmount,
+    address feeAddress
+  );
+
+  constructor(
+    string memory name,
+    string memory symbol,
+    address _paymentProxyAddress
+  ) ERC721(name, symbol) {
+    paymentProxy = _paymentProxyAddress;
+  }
+
+  function supportsInterface(bytes4 interfaceId) public view override(ERC721) returns (bool) {
+    return super.supportsInterface(interfaceId);
+  }
+
+  function payOwner(
+    uint256 receivableTokenId,
+    uint256 amount,
+    bytes calldata paymentReference,
+    uint256 feeAmount,
+    address feeAddress
+  ) external {
+    require(amount != 0, 'Zero amount provided');
+    address owner = ownerOf(receivableTokenId);
+    _paymentId.increment();
+
+    ReceivableInfo storage receivableInfo = receivableInfoMapping[receivableTokenId];
+    address tokenAddress = receivableInfo.tokenAddress;
+
+    (bool status, ) = paymentProxy.delegatecall(
+      abi.encodeWithSignature(
+        'transferFromWithReferenceAndFee(address,address,uint256,bytes,uint256,address)',
+        tokenAddress,
+        owner,
+        amount,
+        paymentReference,
+        feeAmount,
+        feeAddress
+      )
+    );
+    require(status, 'transferFromWithReferenceAndFee failed');
+
+    receivableInfo.balance += amount;
+
+    emit Payment(
+      msg.sender,
+      owner,
+      paymentProxy,
+      receivableTokenId,
+      tokenAddress,
+      _paymentId.current(),
+      paymentReference
+    );
+  }
+
+  function mint(
+    bytes calldata paymentReference,
+    uint256 amount,
+    address erc20Addr,
+    string memory receivableURI
+  ) external {
+    require(paymentReference.length > 0, 'Zero paymentReference provided');
+    require(amount > 0, 'Zero amount provided');
+    require(erc20Addr != address(0), 'Zero address provided');
+    bytes32 idKey = keccak256(abi.encodePacked(msg.sender, paymentReference));
+    require(
+      receivableTokenIdMapping[idKey] == 0,
+      'Receivable has already been minted for this user and request'
+    );
+    _receivableTokenId.increment();
+    uint256 currentReceivableTokenId = _receivableTokenId.current();
+
+    _mint(msg.sender, currentReceivableTokenId);
+    receivableTokenIdMapping[idKey] = currentReceivableTokenId;
+    receivableInfoMapping[currentReceivableTokenId] = ReceivableInfo({
+      tokenAddress: erc20Addr,
+      amount: amount,
+      balance: 0
+    });
```

Comment:
Same as my previous comment - Please make all state changes before interacting with external functions. Even if this code is not susceptible to reentrancy, it's a good practice and may protect future iterations of the code.